### ISSUE TITLE
Weird issue /w missing metrics

### ISSUE THREAD

**Issue by emilebosch** [influxdata/influxdb-ruby#161]
Ok, im running rails 5 /w puma and `influxdb-ruby`. It seems that i'm missing metrics. When testing with apache AB, and the influxdb my metrics do not seem to add up with the count from the test with AB.

For instance.

```
ab -n 1000 http://www.testapp.dev:8000/
```

Doesn't result in 1000 data points. When i run the `influxdb-ruby` gem outside of rails and do a 

```
require 'influxdb'

host = '127.0.0.1'
port = 4444

influxdb = InfluxDB::Client.new udp: { host: host, port: port }

1000.times do |i|
  data = {
    values:    { value: i },
    timestamp: Time.now.to_i
  }
  influxdb.write_point("test", data)
end
```

the data does arrive in the influx db. Could there be some thread locking issue /w puma maybe or something else? I can't get it work So now i resorted to just writing to the plain udp port. Which seems to work

**Comment by dmke** [comment]
I think it may have something to do with the protocol choice of yours. UDP simply has no acknowledge mechanism, so the client can only assume the data has arrived.

The fact that you're using localhost as destination doesn't matter, the OS has only a relatively small cache and you are imposing a huge load with `ab` on the network stack.

You may either try to switch back to HTTP, or try the `async` option (see [README](https://github.com/influxdata/influxdb-ruby#reading-data), scroll up a bit).

**Comment by emilebosch** [comment]
well actually i switched to UDP, because i thought that that the HTTP request might be lagging or just to slow. And i wanted to rule that out. 
So i was going for a fire and forget solution. But it’s really when i include `influx-ruby` that i start missing out metrics. When i just use a plain socket they arrive.

**Comment by dmke** [comment]
Hm... that's indeed weird.

How many points are missing? Which version (gem and server) are you using?

Would you mind sharing (a pointer to) a repo, so I might try to reproduce this? (You can contact me in private at `Base64.decode64("ZG9taW5pay5tZW5rZUBnbWFpbC5jb20=")` if you don't want to share your code publicly.)

**Comment by emilebosch** [comment]
Hey sure ! Il share code and record a screencast tomorrow

**Comment by dmke** [comment]
While playing around, I've found a possible cause for this issue: The default setting for `time_precision` (which defaults to seconds).

If you have setup an `after_action` like this in your controller:

```ruby
def influx_client
  @influx_client ||= InfluxDB::Client.new \
    database:       "whatever",
    host:           "localhost"
end

def log_to_influx
  point = {
    tags: {
      controller: controller_name,
      action:     action_name
    },
    values: {
      code:       self.status,
      remote_ip:  request.remote_ip,
      rand:       rand(1..100)
    },
    timestamp: Time.current.to_i
  }

  influx_client.write_point "responses", point
end
```

you'll end up with **one measurement per second**, i.e. a unique pair `(time, name, tag set)`. In my example, this is `time,whatever,action="$action",controller="$controller"`. If you send multiple points with the same series signature, like this:

```
whatever,action="$action",controller="$controller code-200i,remote_ip="127.0.0.1",rand=12i 1473108595
whatever,action="$action",controller="$controller code-200i,remote_ip="127.0.0.1",rand=8i 1473108595
```

only the last row survives.

To get around this, you'll need to configure at least nano-second precision:

```ruby
  @influx_client ||= InfluxDB::Client.new \
    # ...
    time_presision: "ns"

  point = {
    # ...
    timestamp: (Time.current.to_r * 10**9).to_i
  }
```

If you're unlucky, and two clients hit your app in the same moment (or the OS doesn't provide a precise timer), you could also add the thread id to the tags:

```ruby
    tags: {
      controller: controller_name,
      action:     action_name,
      thread_id:  Thread.current.object_id.to_s(36),
    },
```

(a better choice might be `user_id: current_user.id`, but that would not help unless the user is signed in...)

**Comment by emilebosch** [comment]
Ah right that does seem more like it. Once per second is really too slow, i think ms would be enough for our use case.  However i still don't feel super comfortable with firing off HTTP requests in the dark and having extra logic with it to batch it etc in threads. UDP just seems like such an easier fire and forget mechanism.

This is my current `intializer/metrics.rb` which does the same thing as the gem. 

```
$host = '127.0.0.1'
$port = 4445

$socket = UDPSocket.new

def measure(name, tags, value)
  t = tags.map{|x,y|"#{x}=#{y}"}.join(',')
  d = "#{name},#{t} value=#{value}"
  $socket.send(d, 0, $host, $port)
end


ActiveSupport::Notifications.subscribe "process_action.action_controller" do |name, start, finish, id, payload|
  time = finish - start
  measure "controllers", payload.slice(:controller, :path, :method, :status, :db_runtime, :view_runtime), time
  measure "views", payload.slice(:controller, :status), payload[:db_runtime]
  measure "dbs", payload.slice(:controller, :status), payload[:view_runtime]
end
```

This seems to work pretty well with grafana

**Comment by dmke** [comment]
UDP is a perfectly valid solution.

Just keep in mind, that there's no guarantee of delivery, and the  datagram size (via IPv4) is limited to 65,507 bytes (2**16-1 Bytes - 20 bytes IPv4 header - 8 bytes UDP header).

**Comment by dmke** [comment]
Whoops, wrong button...

**Comment by emilebosch** [comment]
Alright but  think we should implement a Pr anyway because lots of people would be wondering why they are missing rails requests. `ms` would suffice i think. you agree?

**Comment by dmke** [comment]
There's no harm using nano-seconds, since both InfluxDB and Ruby support it (`Time#to_r` is available in [MRI 1.9.3+](http://ruby-doc.org/core-1.9.3/Time.html#method-i-to_r))...

The random noise in the sub-millisecond range might mitigate the uniqueness constraint on the series.

**Comment by emilebosch** [comment]
Thanks for everything, closing this now!
